feat(heureka): adds pagination for component instances on a service details panel

hodanoori opened the pull request:
# Summary
This PR adds pagination to the list of component instances for a selected service on the service details page.

# Changes Made

- Separated the GraphQL query for retrieving the list of component instances from the query for fetching service list data.
- Added a new ComponentInstanceList component to encapsulate the code for displaying component instances and handling its pagination.
- Included ComponentInstancesOfService in the global store slice’s view list to set initial query options, with a default `first: 8` parameter to display a maximum of 8 component instances per page, minimizing the need for scrolling within the panel.
- Renamed "name"to "ccrn" for  "Services," "Components," and "Support Group" entities in all graphql queries.

# Related Issues

- https://github.com/cloudoperators/heureka/issues/337

# Testing Instructions

1. `npm i`
2. `npm run TASK`

# Checklist

<!-- Ensure that your pull request meets the following requirements. -->

- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have made corresponding changes to the documentation (if applicable).
- [x] My changes generate no new warnings or errors.

hodanoori commented:
- closes cloudoperators/heureka#337

barsukov commented on apps/heureka/src/components/services/ComponentInstancesList.jsx:
```diff
@@ -0,0 +1,123 @@
+            onPressNext={() => onPaginationChanged(currentPage + 1)}
+            onPressPrevious={() => onPaginationChanged(currentPage - 1)}
+            onKeyPress={(oKey) => {
+              if (oKey.code === "Enter") {
```
Maybe extract to a naming function here what do you think?

hodanoori replied:
Do you mean to define different functions for each of above functions? I do not see a big advantage of that to be honest. So I leave it as it is, if you don't mind.

barsukov commented on apps/heureka/src/components/services/ComponentInstancesList.jsx:
```diff
@@ -0,0 +1,123 @@
+    enabled: !!queryClientFnReady && !!serviceName,
+  })
+
+  const items = useMemo(() => {
```
I think you probably do not need use memo here do you?

andypf replied:
I think the internal handling of useMemo consumes more performance resources than the benefits it provides in this case.

hodanoori replied:
I remove it, since I think the computation here is not complicate.

barsukov commented on apps/heureka/src/components/services/ComponentInstancesList.jsx:
```diff
@@ -0,0 +1,123 @@
+  }, [data])
+
+  const pageInfo = useMemo(() => {
+    return data?.ComponentInstances?.pageInfo || {}
```
why just do not use directly page info without useMemo?

barsukov reviewed (changes requested):
1) Maybe you do not need useMemo in places, but I am not sure please let me know what do you think 
2) Would be nice to write some tests if it necessary for the components that you introduce.

andypf reviewed (commented):
Great job, Hoda! Please be sure to address the comments.

hodanoori commented:
> 1. Maybe you do not need useMemo in places, but I am not sure please let me know what do you think
> 2. Would be nice to write some tests if it necessary for the components that you introduce.

Thanks for your review. I removed useMemos and added a test for the newly added component. Please have a look here again.

ArtieReus commented on apps/heureka/src/components/services/ComponentInstancesList.jsx:
```diff
@@ -0,0 +1,114 @@
+  const pageInfo = data?.ComponentInstances?.pageInfo || {}
+  const totalPages = pageInfo?.pages?.length || 0
+
+  const onPaginationChanged = (newPage) => {
```
I would recommend to extract this logic to an own component including the Pagination component since you might probably use the pagination in each list and it would give you also the change to tested easier. What do you think?

ArtieReus commented on apps/heureka/src/components/services/ComponentInstancesList.test.jsx:
```diff
@@ -0,0 +1,134 @@
+  })
+
+  // Test: Simulates pagination interaction and verifies query updates
+  it("handles pagination changes", () => {
```
With the component isolation you would just need to test the component and not each list.

ArtieReus reviewed (approved):
lgtm but please take my suggestions into account for the future.

barsukov commented on apps/heureka/src/components/services/ComponentInstancesList.test.jsx:
```diff
@@ -0,0 +1,134 @@
+  default: () => <div>Loading...</div>, // Mock loading element
+}))
+
+vi.mock("../shared/Helper", () => ({
```
Why you also mock it this way?

hodanoori replied:
Cause I did not want to test the severityString, highestSeverity functions from helper inside this component. There is another test for helper.

barsukov reviewed (commented):
I have a few questions regarding mocking, especially if start mocking our own behaviour, it means that the component use this behaviour wrong, kinda a good sign to rethink the state lifecycle.

hodanoori commented on apps/heureka/src/components/services/ComponentInstancesList.test.jsx:
```diff
@@ -0,0 +1,134 @@
+  useGlobalsActions: vi.fn(),
+}))
+
+vi.mock("../shared/LoadElement", () => ({
```
I use this in my component to show loading busy when data isLoading but I did not want to test it here. That's why.

hodanoori commented:
> I have a few questions regarding mocking, especially if start mocking our own behaviour, it means that the component use this behaviour wrong, kinda a good sign to rethink the state lifecycle.

I do not get your point, Those mocks are belonging to other components which has their own tests.

hodanoori commented:
> I have a few questions regarding mocking, especially if start mocking our own behaviour, it means that the component use this behaviour wrong, kinda a good sign to rethink the state lifecycle.

I separated the pagination component and added a test for that component. Please have a look there again.

barsukov reviewed (approved):
LGTM

andypf reviewed (approved):
LGTM
